lib/documents.js

```diff
@@ -2263,12 +2268,184 @@ function writeAllDocumentsImpl(options) {
+       if (!new Set(['forests', 'hosts']).has(String(concurrentRequests.multipleOf))) {
+         throw new Error('Invalid value for onCompletion.multipleOf. Value must be forests or hosts.');
+       }
+     } else {
+       concurrentRequests.multipleOf = 'forests';
+     }
+
+     if (concurrentRequests.multiplier && concurrentRequests.multiplier < 0) {
+       throw new Error('concurrentRequests.multiplier cannot be less than zero');
+     } else {
+       concurrentRequests.multiplier = 4;
+     }
+     jobState.jobOptions.concurrentRequests = concurrentRequests;
+   }
+   else {
+     jobState.jobOptions.concurrentRequests = {multipleOf:'forests', multiplier:4};
+   }
+
+   // Check for jobOptions.batchSize
+   jobState.jobOptions.batchSize = (jobState.jobOptions.batchSize && jobState.jobOptions.batchSize >0) ? jobState.jobOptions.batchSize:100;
+
+   //Check for defaultMetadata
+   jobState.jobOptions.defaultMetadata = (jobState.jobOptions.defaultMetadata)? jobState.jobOptions.defaultMetadata:{};
+
+   //Check for transform
+   jobState.jobOptions.transform = (jobState.jobOptions.transform)? jobState.jobOptions.transform:{};
+ }
+
+  requester.startRequest(operation).result(mlutil.callbackOn(jobState, onWriteAllInit))
+      .catch((err) => operation.stream.emit('error', err));
+
   return inputStream;
 }
 
+function finishWriter(jobState) {
+  jobState.requesterCount--;
+  if(jobState.requesterCount === 0) {
+    if(jobState.jobOptions && jobState.jobOptions.onCompletion){
+      const summary = {
+        docsWrittenSuccessfully: jobState.docsWrittenSuccessfully,
+        docsFailedToBeWritten: jobState.docsFailedToBeWritten,
+        timeElapsed: jobState.timeElapsed
+      };
+      if (jobState.error) {
+        summary.error = jobState.error.toString();
+      }
+      jobState.jobOptions.onCompletion(summary);
+    }
+  }
+}
+
+function onWriteAllInit(output) {
+
+  let maxRequesters = 0;
+  const jobState = this;
+
+  switch (jobState.jobOptions.concurrentRequests.multipleOf) {
+    case 'forests':
+      maxRequesters = jobState.jobOptions.concurrentRequests.multiplier * output.length;
+      break;
+    case 'hosts':
+      maxRequesters = jobState.jobOptions.concurrentRequests.multiplier *
+          new Set(output.map((forest) => forest.host)).size;
+      break;
+    default:
+      throw new Error('Invalid value for onCompletion.multipleOf. Value must be forests or hosts.');
+  }
+  jobState.startTime = Date.now();
+  for (let i = 0; i < maxRequesters; i++) {
+    jobState.requesterCount++;
+    onWriteAllDocs(jobState,i);
+  }
+}
+
+function onWriteAllDocs(jobState, writerId) {
+
+  const writeBatchArray = [];
+  const batchSize = jobState.jobOptions.batchSize;
+  jobState.lastBatch = false;
+  jobState.retryCount = 2;
+
+  a:
+      for (let i = 0; i < batchSize; i++) {
+        const record = jobState.stream.read();
+
+        if(record !== null) {
+          if (jobState.jobOptions.defaultMetadata.length>0) {
```

Comment:
Should this test whether defaultMetadata is set?

Because defaultMetadata is an object instead of an array, it shouldn't have a length property. 

Also, the test should check i === 0 so defaultMetadata is pushed only before the first document.  At present, the default metadata object is being added before each document, which adds to the cost with the same functionality.

Reply:
Changed the same for "transform" as well. Should transform also be added only if i===0?